[FC-0040] feat: update search index when object tags change

yusuf-musleh opened the pull request:
## Description

This PR implements updating search index with object tags whenever they change, based on the new `CONTENT_OBJECT_TAGS_CHANGED` event. 

## Supporting Information
- Related Ticket: https://github.com/openedx/modular-learning/issues/197
- Depends on: https://github.com/openedx/openedx-events/pull/327

## Testing Instructions
- Run your local devstack on this branch
- Clone the openedx-events repo on this branch https://github.com/openedx/openedx-events/pull/327 and make sure it is accessible from your devstack, and install it locally
  - `pip uninstall openedx-events`
  - `pip install -e path/to/openedx-events/`
- Make sure you have `meilisearch` setup locally, follow the setup instructions here https://github.com/open-craft/tutor-contrib-meilisearch
- Make sure you also have some sample taxonomies/tags, you can add some from here: https://github.com/open-craft/taxonomy-sample-data
- Run `tutor dev run cms bash` and `./manage.py cms reindex_studio`
- View the resulting index at http://meilisearch.local.edly.io:7700/ (see [tutor-contrib-meilisearch README](https://github.com/open-craft/tutor-contrib-meilisearch) for how to get the API key to log in)
- Check and confirm that is an index with documents in it. If you already have some tags applied to blocks, they will appear there
- Go back to studio, and add/update tags in a block
- Go back to meilisearch dashboard, and confirm that the `Tags` field in the index document for the block has been updated to reflect that changes you made.
  - Note: I recommend having a easily identifiable name for the block/unit that you added tags to so you can easily search for it
  - **NOTE:** Meilisearch seems to cache queries along with their results in the frontend, so if you simply search it might show you stale data (network tab shows no requests), especially if you're searching for the same query. Make sure you refresh the Meilisearch dashboard (http://meilisearch.local.edly.io:7700/) and then perform the search.
- Repeat and confirm the above for both units and blocks

---
Prinvate-ref: [FAL-3691](https://tasks.opencraft.com/browse/FAL-3691)

feanil committed:
fix: Remove deprecated getargspec call.

This function was removed by python 3.11 so update to the alternate
call that is the current recommended replacement.

https://docs.python.org/3.11/library/inspect.html#inspect.getfullargspec

feanil committed:
fix: Don't use the deprecated location for Hashable

The Hashable object was moved in python 3.3 and support for the old
location is dropped in python 3.10 the new location is available in
python 3.8 so we can just update this and it should work with both
python 3.8 and 3.11

https://docs.python.org/3.8/library/collections.html

feanil committed:
fix: Fix function mocking.

The way the patch decorator was being used is not supported in python
3.11.  Use the patch decorator to auto generate the correct mock and
make the test a bit more readabale.  The new change is both 3.8 and
3.11 compatible.

feanil committed:
fix: Create a bad unicode file differently.

In Python 3.11 CSV files are allowed to have null characters so the test
data is no longer a valid. We update it to not have a valid unicode
character to still test this code path correctly.

I'm not concerned about the fact that the files with null will get past
this test beacause there are other checks on the content of the file
that catch if it doesn't have enough or the right fields so this should
be a safe change to make to the tests.

Relevant Change in Python: https://github.com/python/cpython/issues/71767

feanil committed:
fix: Provide a sequence to random.sample

The sample function used to automatically convert sets to sequences but
that is no longer supported starting in 3.11 so we have to do it
manually.

Reference: https://docs.python.org/3/library/random.html#random.sample

yusuf-musleh committed:
feat: Update search index when object tags updated

yusuf-musleh commented on openedx/core/djangoapps/content/search/handlers.py:
```diff
@@ -68,3 +69,22 @@ def xblock_deleted_handler(**kwargs) -> None:
+        log.error("Received null or incorrect data for event")
+        return
+
+    upsert_xblock_index_doc.delay(
+        content_object_tags.object_id,
+        recursive=False,
+        update_metadata=True,
+        update_tags=True,
+    )
```
I needed to set `update_metadata=True` here, because without it, it was creating new documents when tags are updated, rather than updating the existing one.

rpenido replied:
I think the new documents' issue is unrelated to this parameter.

We changed the way the ID hash was generated upstream, so the first time you tag an object (after the update), a new document is generated. Could you check if this was the case?

If so, using `update_metadata=False` should work.

rpenido replied:
Also, adding Tags to a LibraryComponent if passing here (and falling).

A suggestion is to refactor and create a new update function for tagging only that can handle both types of blocks calling `searchable_doc_for_library_block` or `searchable_doc_for_course_block` (and eliminate the `update_metadata` and `update_tags` parameters from the task). But it's up to you!

yusuf-musleh replied:
> Could you check if this was the case?
> If so, using update_metadata=False should work.

Yes -- that seemed to fix it and using `update_metadata=False` now works! I updated it.

I also added handling tagging for library blocks, I just opted for checking the type of the object_id for the content object and choosing the correct method based on that.

yusuf-musleh commented on openedx/core/djangoapps/content/search/documents.py:
```diff
@@ -162,7 +162,7 @@ def _tags_for_content_object(object_id: UsageKey | LearningContextKey) -> dict:
     all_tags = tagging_api.get_object_tags(object_id).all()
     if not all_tags:
-        return {}
+        return {Fields.tags: {}}
```
This was needed to clear out tags in the index when unselecting all tags for the block, it would remain the last value otherwise because the `Fields.tags` value would not be included.

bradenmacdonald replied:
Might be worth mentioning that in the code as a comment :)

yusuf-musleh commented on openedx/core/djangoapps/content/search/tests/test_views.py:
```diff
@@ -1,83 +1,83 @@
-Tests for the Studio content search REST API.
-"""
-from __future__ import annotations
+# """
```
Commented out tests in this file, so other tests can run, since this will be fixed upstream.

rpenido commented on openedx/core/djangoapps/content/search/documents.py:
```diff
@@ -230,6 +230,7 @@ def searchable_doc_for_course_block(block, metadata: bool = True, tags: bool = T
     doc = {
         Fields.id: meili_id_from_opaque_key(block.usage_key),
         Fields.type: DocType.course_block,
+        Fields.tags: {},
```
This line will clear the tags if we update an XBlock. Maybe the change in `_tags_for_content_object()` already fixed it? (I removed this line while testing and it seems to work)

```suggestion
```

yusuf-musleh replied:
You're right, I added this mainly so all the docs have the same format that has (containing `tags` field, whether or not there are tags), however this caused the issue you mentioned, so I removed it.

rpenido reviewed (commented):
Hi @yusuf-musleh! Good work here!
I left some suggestions and questions in the review.

mariajgrimaldi commented on openedx/core/djangoapps/content_tagging/rest_api/v1/views.py:
```diff
@@ -143,12 +146,26 @@ def orgs(self, request, **_kwargs) -> Response:
+        if response.status_code == 200:
+            object_id = kwargs.get('object_id')
+            CONTENT_OBJECT_TAGS_CHANGED.send_event(
+                time=now(),
```
The event implemented [here](https://github.com/openedx/openedx-events/pull/327) explicitly sends only content_object, is this attribute missing here?

yusuf-musleh replied:
When I was implementing this, I thought this was needed as part of sending events based on other events in the codebase, but you're right, its actually extra and not needed, I removed it. Thanks for catching this!

mariajgrimaldi commented on openedx/core/djangoapps/content_tagging/rest_api/v1/views.py:
```diff
@@ -143,12 +146,26 @@ def orgs(self, request, **_kwargs) -> Response:
+            object_id = kwargs.get('object_id')
+            CONTENT_OBJECT_TAGS_CHANGED.send_event(
+                time=now(),
+                content_object=ContentObjectData(object_id=object_id)
```
Also I wonder, what's an `object_id`? What does it look like?

yusuf-musleh replied:
The `object_id` is the id of the course or library block as a string that is being tagged. For example:

`block-v1:SampleTaxonomyOrg2+STC1+2023_1+type@vertical+block@f8de78f0897049ce997777a3a31b6ea0`

mariajgrimaldi replied:
Can we add that explanation to the event data docstring? [Here](https://github.com/openedx/openedx-events/pull/327/files#diff-ce3d23f91ef18fd93f59fd1cf0a2c5f88b98dfde3f3790cb22281e74edfad042R191) that would address all my comments. Thank you!

yusuf-musleh replied:
Sure thing! Updated.

mariajgrimaldi reviewed (changes requested):
I left a few comments regarding https://github.com/openedx/openedx-events/pull/327. Also, is this PR going to be open against upstream?

yusuf-musleh committed:
feat: Update index when library block tags change

yusuf-musleh committed:
refactor: Remove extra params

yusuf-musleh commented:
Thanks @rpenido and @mariajgrimaldi for your reviews! I've addressed your comments.

> Also, is this PR going to be open against upstream?

Yes, this PR should be opened against upstream master once the base of this PR gets merged.

rpenido reviewed (approved):
Tested again and is working good @yusuf-musleh! :+1:

- [x] I tested this using the Testing Instructions
- [x] I read through the code
- [x] ~~I checked for accessibility issues~~
- [x] Includes documentation

One nit:
We have a list of the current `edx-platform` events here:
https://github.com/openedx/edx-platform/blob/master/docs/hooks/events.rst

Could you add this new event to the list? Thank you!

yusuf-musleh committed:
docs: Add new event to hook events docs

yusuf-musleh commented:
> Could you add this new event to the list? Thank you!

@rpenido Sure thing, added!

feanil committed:
fix: Be able to clear the process_cache manually in Python 3.11

Given code like the following

```
class Foo:
    @process_cached
    def bar(self):
        pass
```

In Python 3.8 referencing `bar` would not call its `__get__` method.

```
x = Foo().bar
```

However in Python 3.11, making the same call would call the `__get__`
method, permanently replacing the underlying `process_cached` object
with the partial function that references it.

This meant that code to clear the cache would work in Python 3.8 but
would break in 3.11

```
Foo().bar.cache.clear()  # Works in 3.8 but not in 3.11
```

In 3.11 this results in the following error:
```
E       AttributeError: 'functools.partial' object has no attribute 'cache'
```

To make this compatible in both version, we just add the cache as an
accessible attribute on the partial we generate for our wrapped
function.

bradenmacdonald commented:
@yusuf-musleh What's the status of this PR - are you waiting for a review from me?

yusuf-musleh commented:
@bradenmacdonald I was waiting on upstream https://github.com/openedx/edx-platform/pull/34391 to be merged first so I can rebase and fix any issues before requesting your review, since I would need to close this PR and open a new on against master at that point.

bradenmacdonald commented:
@yusuf-musleh OK. Unfortunately there are some issues on edx-platform that are preventing us from merging any PRs until next week (CC @rpenido).

But that PR is approved and ready to go so you can rebase on it.

yusuf-musleh committed:
feat: Add and use upsert_block_tags_index_docs

This takes care of updating tags data in search index for both course and library blocks.

yusuf-musleh committed:
chore: Update openedx-events

yusuf-musleh committed:
fix: Update tests + include course block tags in reindex

yusuf-musleh committed:
feat: Fix static-type issues + adjust tag_object

bradenmacdonald commented:
Thanks @yusuf-musleh! The code looks good. Just waiting until we can merge that other PR; I don't expect this PR to need any further changes.

kdmccormick committed:
refactor: remove requirements/edx-sandbox/shared.[in,txt]

These files were used to assist the Python 3.5 -> 3.8 upgrade,
but they are no longer needed nor referened anywhere. They haven't
been updated for years.

kdmccormick committed:
feat!: expose per-release edx-sandbox dependency pins

See requirements/edx-sandbox/README.rst for more info

BREAKING CHANGE: edx-sandbox/py38.txt will not longer
be updated. Please install from either edx-sandbox/base.txt or
edx-sandbox/releases/*.txt instead.

kdmccormick committed:
feat: link py38.txt->release/quince.txt for backwards compatibility

kdmccormick committed:
build: replace wget->curl, so make upgrade works in tutor

tutor's containers don't have wget installed, and curl -L works
just as well and is installed into basically everything

kdmccormick committed:
feat: freeze edx-sandbox requirements for quince

We ran:

    cp requirements/edx-sandbox/base.txt \
       requirements/edx-sandbox/releases/quince.txt

feanil committed:
feat!: Remove the django-splash app.

DEPR: https://github.com/openedx/public-engineering/issues/224

The django-splash repo was created 11 years ago to let the LMS redirect
users to a splash screen when a user comes to the site for the first
time. It works by looking for a configurable cookie value and
redirecting from the middleware.

This feature was never documented, has some edx.org hardcoded defaults,
and is not compatible with MFEs.

BREAKING CHANGE: The django splash feature will no longer be available.

feanil committed:
chore: Run `make compile-requirements` to drop django-splash.

feanil committed:
test: Reduce query counts now that we dropped django-splash.

feanil committed:
feat: Upgrade Python dependency edx-bulk-grades

Update to a Python 3.11 compatible version

Commit generated by workflow `openedx/edx-platform/.github/workflows/upgrade-one-python-dependency.yml@refs/heads/master`

jristau1984 committed:
chore: update 2U teams in CODEOWNERS (#34352)

feanil committed:
feat: Upgrade Python dependency edx-django-sites-extensions

Update to a Python 3.11 compatible version

Commit generated by workflow `openedx/edx-platform/.github/workflows/upgrade-one-python-dependency.yml@refs/heads/master`

rpenido committed:
feat: update Studio search index when course content is updated (#34391)

Commit:
Merge pull request #34533 from openedx/feanil/upgrade-edx-bulk-grades-890aa15

feat: Upgrade Python dependency edx-bulk-grades

feanil committed:
feat: Upgrade Python dependency edx-drf-extensions

Update to a Python 3.11 compatible version

Commit generated by workflow `openedx/edx-platform/.github/workflows/upgrade-one-python-dependency.yml@refs/heads/master`

connorhaugh committed:
temp: trace celery tasks in dd (#34537)

feanil committed:
feat: Upgrade Python dependency edx-milestones

Update to a Python 3.11 compatible version

Commit generated by workflow `openedx/edx-platform/.github/workflows/upgrade-one-python-dependency.yml@refs/heads/master`

Commit:
Merge pull request #34541 from openedx/feanil/upgrade-edx-django-sites-extensions-f18629e

feat: Upgrade Python dependency edx-django-sites-extensions

Commit:
Merge pull request #34542 from openedx/feanil/upgrade-edx-drf-extensions-16b82db

feat: Upgrade Python dependency edx-drf-extensions

Commit:
chore: version bump

katrinan029 committed:
Merge pull request #34544 from openedx/knguyen2/version-bump

chore: version bump

Commit:
Merge pull request #34543 from openedx/feanil/upgrade-edx-milestones-16b82db

feat: Upgrade Python dependency edx-milestones

hamzawaleed01 committed:
feat: Upgrade Python dependency edx-enterprise

fix: SSO self-serve tool invalid entityId parsing

Commit generated by workflow `openedx/edx-platform/.github/workflows/upgrade-one-python-dependency.yml@refs/heads/master`

Commit:
Merge pull request #34506 from openedx/feanil/fix_tests_for_python_3.11

Make Tests Python 3.11 Compatible

hamzawaleed01 committed:
Merge pull request #34546 from openedx/hamzawaleed01/upgrade-edx-enterprise-68b3753

feat: Upgrade Python dependency edx-enterprise

Commit:
Merge pull request #34528 from openedx/feanil/remove_django_splash

feat!: Remove the django-splash app.

github-actions[bot] committed:
feat: Upgrade Python dependency Xblock (#34547)

feat: re-expose deprecated fragments as passthrough

Commit generated by workflow `openedx/edx-platform/.github/workflows/upgrade-one-python-dependency.yml@refs/heads/master`

Co-authored-by: connorhaugh <connorhaugh@users.noreply.github.com>
Co-authored-by: connorhaugh <49422820+connorhaugh@users.noreply.github.com>

feanil committed:
feat: Upgrade Python dependency edx-proctoring

Update to a Python 3.11 compatible version

Commit generated by workflow `openedx/edx-platform/.github/workflows/upgrade-one-python-dependency.yml@refs/heads/master`

feanil committed:
feat: Upgrade Python dependency edx-rbac

Update to a Python 3.11 compatible version

Commit generated by workflow `openedx/edx-platform/.github/workflows/upgrade-one-python-dependency.yml@refs/heads/master`

bradenmacdonald committed:
fix: bug retrieving the Meilisearch API key UID (first time only)

bradenmacdonald commented on openedx/core/djangoapps/content/search/tests/test_api.py:
```diff
@@ -116,11 +129,19 @@ def test_reindex_meilisearch_disabled(self, mock_meilisearch):
     @override_settings(MEILISEARCH_ENABLED=True)
     def test_reindex_meilisearch(self, mock_meilisearch):
 
+        # Add tags field to doc, since reindex calls with `include_tags=True` internally
```
This docstring is out of date - we don't have `include_tags=True` anymore

connorhaugh committed:
Revert "temp: trace celery tasks in dd (#34537)" (#34553)

This reverts commit f866545bb9a48ffe4e5ef0402fdc9819c4ed6f0c.

Commit:
Merge pull request #34551 from openedx/feanil/upgrade-edx-proctoring-1eb92a3

feat: Upgrade Python dependency edx-proctoring

Commit:
Merge pull request #34552 from openedx/feanil/upgrade-edx-rbac-1eb92a3

feat: Upgrade Python dependency edx-rbac

yusuf-musleh committed:
docs: Update comments

yusuf-musleh commented:
> I found a small but significant bug, so I pushed a fix to your PR. And (unrelated) there are some conflicts now. Could you please rebase or merge master into this, then ping me to merge it?

@bradenmacdonald Got it, I've addressed your comments, rebased/fixed conflicts and created a new PR off master, thats good to go (https://github.com/openedx/edx-platform/pull/34559). So I'm closing this PR in favor of that one.
